File: bst/java/LinkedListStack.java
public class LinkedListStack<Item> {
  private Node first;

  private class Node {
    Item item;
    Node next;

    Node(Item item, Node next) {
      this.item = item;
      this.next = next;
    }
  }

  public boolean isEmpty() {
    return first == null;
  }

  public void push(Item item) {
    first = new Node(item, first);
  }

  public Item pop() {
    if(isEmpty()) return null;
    Item item = first.item;
    first =  first.next;
    return item;
  }
}

File: stack/ruby/lib/stack/linked_list_stack.rb
class Stack::LinkedListStack
  include Enumerable
  class Node
    attr_accessor :next, :item

    def initialize(item, nextNode)
      self.next = nextNode
      self.item = item
    end
  end

  def isEmpty?
    @first.nil?
  end

  def push(item)
    @first = Node.new(item, @first)
  end

  def pop
    return if isEmpty?
    item = @first.item
    @first = @first.next
    item
  end

  def each(&block)
    current = @first
    while(current) do
      block.call(current.item)
      current = current.next
    end
  end
end
